src/GitVersionTask/GitVersionTaskBase.cs

```diff
@@ -1,72 +1,141 @@
 namespace GitVersionTask
 {
-    using System;
     using GitVersion;
     using GitVersion.Helpers;
+    using System;
+    using System.IO;
 
-    using Microsoft.Build.Framework;
-    using Microsoft.Build.Utilities;
-
-    public abstract class GitVersionTaskBase : Task
+    public static class GitVersionTaskBase
     {
-        protected GitVersionTaskBase()
+        internal static TOutput ExecuteGitVersionTask<TInput, TOutput>(
+            TInput input,
+            Func<TInput, TaskLogger, TOutput> execute
+            )
+            where TInput : AbstractInput
+            where TOutput : class, new()
         {
-            var fileSystem = new FileSystem();
-            ExecuteCore = new ExecuteCore(fileSystem);
-            GitVersion.Logger.SetLoggers(LogDebug, LogInfo, LogWarning, s => LogError(s));
-        }
+            if (!input.ValidateInput())
+            {
+                throw new Exception("Invalid input.");
```

Comment:
Don't we have a more specific exception type to throw here? Also, wouldn't it be better if `input.ValidateInput()` threw the exception, as it would yield a deeper and more informative stack trace?